### PR TITLE
Add Network Components

### PR DESCRIPTION
**Pull request by michavie**
Migration from https://github.com/PeerMeHQ/mx-sdk-php-network-providers

Dependencies:
- `guzzlehttp/guzzle`:  [Repository](https://github.com/guzzle/guzzle) | [Distribution](https://packagist.org/packages/guzzlehttp/guzzle) | Installs: 633m (Http client lib)
- `nesbot/carbon`: [Repository](https://github.com/briannesbitt/Carbon) | [Distribution](https://packagist.org/packages/nesbot/carbon) | Installs: 407m (popular & commonly used datetime wrapper)
- `spatie/phpunit-snapshot-assertions`: [Repository](https://github.com/spatie/phpunit-snapshot-assertions) | [Distribution](https://packagist.org/packages/spatie/phpunit-snapshot-assertions) | Installs: 7m (Snapshot assertions maintained by Spatie company, known for maintaining their packages)

### CODE REVIEW

File: src/Http/Api/Endpoints/VmEndpoints.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    public function query(string $contractAddress, string $func, array $args = [], array $params = []): VmQueryResult
+    {
+        return VmQueryResult::fromApiResponse(
+            $this->client->request('POST', "/vm-values/query", [
+                'json' => [
+                    'scAddress' => $contractAddress,
+                    'funcName' => $func,
+                    'args' => collect($args)->map(fn ($a) => Encoder::toHex($a))->all(),
```
Comment (hschiau) [inline src/Http/Api/Endpoints/VmEndpoints.php]: What do you think about making the encoding optional? Give the user the choice over handling it themselves. We could add an extra parameter for this

File: src/Http/Entities/IEntity.php [hunk 2]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+namespace MultiversX\Http\Entities;
+
+use Illuminate\Support\Collection;
+use Psr\Http\Message\ResponseInterface;
+
+interface IEntity
+{
+    public static function fromApiResponse(ResponseInterface $res, bool $isCollection = false): IEntity|Collection;
```
Comment (hschiau) [inline src/Http/Entities/IEntity.php]: I noticed the function signature for the implementation inside the `HasApiResponses` trait is slightly different. It contains an additional parameter (`bool $unwrapData = false`). Could this cause any issues ?

File: composer.json [hunk 3]
```diff
@@ -22,10 +22,13 @@
     "bitwasp/bech32": "^0.0.1",
     "brick/math": "^0.11",
     "kornrunner/keccak": "^1.1",
-    "illuminate/collections": "^9.0 || ^10.0"
+    "illuminate/collections": "^9.0 || ^10.0",
+    "guzzlehttp/guzzle": "^7.5",
+    "nesbot/carbon": "^2.66"
```
Comment (andreibancioiu) [inline composer.json:27]: As previously, perhaps add a few GitHub details in the PR description?

File: src/Http/Api/Exceptions/HttpClientException.php [hunk 4]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace MultiversX\Http\Exceptions;
+
+use Exception;
+
+class HttpClientException extends Exception
+{
+    //
```
Comment (andreibancioiu) [inline src/Http/Api/Exceptions/HttpClientException.php]: Perhaps drop comment - or leave it as it is, if that's more common for empty classes in PHP.

File: src/Http/Api/HasApiResponses.php [hunk 5]
```diff
@@ -0,0 +1,68 @@
+<?php
+
+namespace MultiversX\Http\Api;
+
+use Illuminate\Support\Collection;
+use MultiversX\Http\Entities\IEntity;
+use MultiversX\Http\Exceptions\RequestException;
+use Psr\Http\Message\ResponseInterface;
+
+trait HasApiResponses
```
Comment (andreibancioiu) [inline src/Http/Api/HasApiResponses.php:10]: Nice, traits in PHP :rocket:

**Review by andreibancioiu — DISMISSED** [review]
Great work :rocket: 

When sdk-specs for network providers are ready (they will depict a minimalistic required portion, with freedom for implementing the rest of API / Gateway features), some subsequent adjustments might be applied.

The unit tests are quite extensive.

File: composer.json (hunk 3 above)
Comment (michavie) [reply composer.json:27]: done ✅

File: src/Http/Api/Exceptions/HttpClientException.php (hunk 4 above)
Comment (michavie) [reply src/Http/Api/Exceptions/HttpClientException.php]: Removed✅

File: src/Http/Api/Endpoints/VmEndpoints.php (hunk 1 above)
Comment (michavie) [reply src/Http/Api/Endpoints/VmEndpoints.php]: Good point; added an optional param as you proposed. Would default it to true (auto-encode), as in probably +95% of cases it is the wanted behavior

File: src/Http/Entities/IEntity.php (hunk 2 above)
Comment (michavie) [reply src/Http/Entities/IEntity.php]: Good catch! Fixed ✅

**Comment by michavie** [comment]
@andreibancioiu pushed updates

Regarding unit tests, I added snapshot validations for each endpoint to ensure correctness of response & type given the fragility – that's the intent behind it. Let me know if you see overlaps or think some should be purged.

**Comment by michavie** [comment]
Also added a test workflow in #4 . With network providers + testing merged, we can go ahead and release a first version

**Review by andreibancioiu — APPROVED** [review]

**Comment by andreibancioiu** [comment]
> @andreibancioiu pushed updates
> 
> Regarding unit tests, I added snapshot validations for each endpoint to ensure correctness of response & type given the fragility – that's the intent behind it. Let me know if you see overlaps or think some should be purged.

It's great - we will brainstorm a bit whether we should ask test vectors in PY & JS, as well :pray:

**Review by hschiau — APPROVED** [review]

**Pull request merged**